# Artificial Neural Network

### Importing the libraries

In [1]:
import numpy as np
import pandas as pd
import tensorflow as tf

to show the latest version of the tensorflow on Google Colab

In [2]:
tf.__version__

'2.15.0'

## Part 1 - Data Preprocessing

### Importing the dataset

In [4]:
dataset = pd.read_csv('Churn_Modelling.csv')
# the first three columns don't have any impact on a customer loyalty
X = dataset.iloc[:, 3:-1].values
y = dataset.iloc[:, -1].values

In [5]:
print(X)

[[619 'France' 'Female' ... 1 1 101348.88]
 [608 'Spain' 'Female' ... 0 1 112542.58]
 [502 'France' 'Female' ... 1 0 113931.57]
 ...
 [709 'France' 'Female' ... 0 1 42085.58]
 [772 'Germany' 'Male' ... 1 0 92888.52]
 [792 'France' 'Female' ... 1 0 38190.78]]


In [6]:
print(y)

[1 0 1 ... 1 1 0]


### Encoding categorical data

Label Encoding the "Gender" column (categorical data)

In [7]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
# the third column of X not the actual csv data is Gender --> index is 2
X[:, 2] = le.fit_transform(X[:, 2])

In [8]:
print(X)

[[619 'France' 0 ... 1 1 101348.88]
 [608 'Spain' 0 ... 0 1 112542.58]
 [502 'France' 0 ... 1 0 113931.57]
 ...
 [709 'France' 0 ... 0 1 42085.58]
 [772 'Germany' 1 ... 1 0 92888.52]
 [792 'France' 0 ... 1 0 38190.78]]


One Hot Encoding the "Geography" column (categorical data)

here we use One Hot Encoding for three countries since they don't have any

order

In [9]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
# the second column of X is for Geography --> index = 1
ct = ColumnTransformer(transformers= [('encoder', OneHotEncoder(), [1])], remainder = 'passthrough')
X = np.array(ct.fit_transform(X))

In [10]:
print(X)

[[1.0 0.0 0.0 ... 1 1 101348.88]
 [0.0 0.0 1.0 ... 0 1 112542.58]
 [1.0 0.0 0.0 ... 1 0 113931.57]
 ...
 [1.0 0.0 0.0 ... 0 1 42085.58]
 [0.0 1.0 0.0 ... 1 0 92888.52]
 [1.0 0.0 0.0 ... 1 0 38190.78]]


### Splitting the dataset into the Training set and Test set

In [11]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split (X, y, test_size = 0.2, random_state = 0)

### Feature Scaling

Apply feature scaling for both training and test set

In [12]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

## Part 2 - Building the ANN

### Initializing the ANN

Artificial Neural Network is a sequence of layers

In [13]:
ann = tf.keras.models.Sequential()

### Adding the input layer and the first hidden layer

Dense is a fully connected layer, number of neurons in the hidden layers = units = 6

There is no rule of thumbs for the number of hidden layers

activation functions = activation = rectifier or 'relu'

In [14]:
ann.add(tf.keras.layers.Dense(units=6, activation='relu'))

### Adding the second hidden layer

In [16]:
ann.add(tf.keras.layers.Dense(units=6, activation='relu'))

### Adding the output layer

sigmoid is giving us a binary result

In [17]:
ann.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))

## Part 3 - Training the ANN

### Compiling the ANN

Best optimizer is Gradient Descent here 'adam' optimizer

loss function here is binary so --> binary_crossentropy

if the loss function is not binary we have to choose another one (also change activation function)

for the metrics we choose accuracy

In [18]:
ann.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['accuracy'])

### Training the ANN on the Training set

here we use batch learning, and by default batch_size = 32

In [19]:
ann.fit(X_train, y_train, batch_size = 32, epochs = 100)

Epoch 1/100
250/250 [==============================] - 2s 2ms/step - loss: 0.5946 - accuracy: 0.7174
Epoch 2/100
250/250 [==============================] - 1s 4ms/step - loss: 0.4541 - accuracy: 0.8001
Epoch 3/100
250/250 [==============================] - 1s 4ms/step - loss: 0.4288 - accuracy: 0.8062
Epoch 4/100
250/250 [==============================] - 1s 5ms/step - loss: 0.4185 - accuracy: 0.8096
Epoch 5/100
250/250 [==============================] - 1s 3ms/step - loss: 0.4116 - accuracy: 0.8171
Epoch 6/100
250/250 [==============================] - 1s 2ms/step - loss: 0.4034 - accuracy: 0.8230
Epoch 7/100
250/250 [==============================] - 1s 2ms/step - loss: 0.3947 - accuracy: 0.8315
Epoch 8/100
250/250 [==============================] - 1s 3ms/step - loss: 0.3856 - accuracy: 0.8382
Epoch 9/100
250/250 [==============================] - 1s 2ms/step - loss: 0.3769 - accuracy: 0.8459
Epoch 10/100
250/250 [==============================] - 1s 2ms/step - loss: 0.3677 - accura

The first result: loss: 0.5946 - accuracy: 0.7174

The 100th result: loss: 0.3260 - accuracy: 0.8669

both loss function and accuracy are improved

## Part 4 - Making the predictions and evaluating the model

### Predicting the result of a single observation

**Homework**

Use our ANN model to predict if the customer with the following informations will leave the bank:

Geography: France

Credit Score: 600

Gender: Male

Age: 40 years old

Tenure: 3 years

Balance: \$ 60000

Number of Products: 2

Does this customer have a credit card? Yes

Is this customer an Active Member: Yes

Estimated Salary: \$ 50000

So, should we say goodbye to that customer?

**Solution**

France is one hot encoded to dummy variable 100 here: 1, 0, 0

Male is label encoded to 1 and female to 0

in the sigmoid function: if the result > 0.5 --> result =1, else --> result = 0

In [20]:
print(ann.predict(sc.transform([[1, 0, 0, 600, 1, 40, 3, 60000, 2, 1, 1, 50000]])))

1/1 [==============================] - 1s 601ms/step
[[0.02326598]]


In [21]:
print(ann.predict(sc.transform([[1, 0, 0, 600, 1, 40, 3, 60000, 2, 1, 1, 50000]])) > 0.5)

1/1 [==============================] - 0s 26ms/step
[[False]]


Therefore, our ANN model predicts that this customer stays in the bank!

**Important note 1:** Notice that the values of the features were all input in a double pair of square brackets. That's because the "predict" method always expects a 2D array as the format of its inputs. And putting our values into a double pair of square brackets makes the input exactly a 2D array.

**Important note 2:** Notice also that the "France" country was not input as a string in the last column but as "1, 0, 0" in the first three columns. That's because of course the predict method expects the one-hot-encoded values of the state, and as we see in the first row of the matrix of features X, "France" was encoded as "1, 0, 0". And be careful to include these values in the first three columns, because the dummy variables are always created in the first columns.

### Predicting the Test set results

In [22]:
y_pred = ann.predict(X_test)
y_pred = (y_pred > 0.5)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1))

63/63 [==============================] - 0s 2ms/step
[[0 0]
 [0 1]
 [0 0]
 ...
 [0 0]
 [0 0]
 [0 0]]


### Making the Confusion Matrix

In [24]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)

[[1511   84]
 [ 193  212]]


0.8615